### ISSUE TITLE
Improvements to arc-eager and arc-standard algorithms

### ISSUE THREAD

**Issue by stevenbird** [nltk/nltk#905]
Joakim Nivre has made some improvements to both of these algorithms. @longdt219 please show him the implementations you have created and request his input.

**Comment by longdt219** [comment]
Hi @stevenbird, 
How do you know that he has made some improvement ? Should I email him asking for these changes ?

**Comment by stevenbird** [comment]
Yes. Please include a pointer to the implementations you've already done.

**Comment by stevenbird** [comment]
Joakim Nivre recommends the following improvements:
1. Add the Swap transition to the arc-standard parser in order to allow non-projective trees. This should be relatively straightforward.
2. Add a dynamic oracle for the arc-eager parser. This requires a switch from batch learning to online learning.
3. Add beam search and structured learning à la Zhang and Clark (2008) to both parsers. This requires reimplementing both the learning and parsing procedures.

**Comment by stevenbird** [comment]
@longdt219 please go ahead with the swap transition, and illustrate it using Czech (depends on https://github.com/nltk/nltk/issues/875)

**Comment by no-identd** [comment]
>depends on #875

and on https://github.com/nltk/nltk_data/issues/20, right?
